torch/csrc/monitor/instrumentation.h

```diff
@@ -7,12 +7,35 @@
 
 #include <c10/macros/Macros.h>
 #include <c10/util/ScopeExit.h>
+#include <c10/util/SmallVector.h>
 
-namespace torch {
-namespace monitor {
+namespace torch::monitor {
 namespace detail {
 class WaitCounterImpl;
-}
+
+class WaitCounterBackendIf {
+ public:
+  virtual ~WaitCounterBackendIf() = default;
+
+  virtual intptr_t start(
+      std::chrono::steady_clock::time_point now) noexcept = 0;
+  virtual void stop(
+      std::chrono::steady_clock::time_point now,
+      intptr_t ctx) noexcept = 0;
```

Comment:
Not sure if this is a part of another diff but I do not understand why do we pass ctx everywhere?

Reply:
This enables backend to carry certain state from start to stop.
E.g. in the most basic case of measuring the duration of operation, this context can simply contain start time.

But if the backend decides to e.g. start an OTEL span for every start/stop pair (using example from https://opentelemetry.io/docs/languages/cpp/instrumentation/#start-a-span) then:
1. start() will call StartSpan (and potentially WithActiveSpan if scope is needed too). The span object will be converted into a context (e.g. by wrapping it into a unique_ptr, then doing .release() + reinterpret_cast)
2. stop() will cast ctx back into the unique ptr and destroy the span/scope.